### PR TITLE
backup: Allow for upgrade safe backups

### LINKED ISSUE #13448: Feature Request: Allow for backups to be created that are safe to upgrade MySQL with
### Feature Description

In `vtbackup` it's possible to use the `-restart_before_backup` flag to force a restart of MySQL with `innodb_fast_shutdown=0`. This way the backup that is created is safe to also use for upgrading MySQL. When upgrading MySQL, it's necessary to not use fast shutdown as the new version might have a different on disk format.

In `Backup` / `BackupShard` commands though through `vtctld`, it's not possible to specify a similar flag. This means that backups created through that method couldn't be safely used for upgrades. 

Request here is to have a flag available for those backup methods as well so it's possible to create a safe to upgrade backup.

### Use Case(s)

Useful when upgrading MySQL in a Vitess cluster, especially when doing it in an automated fashion for a large number of databases.

### PR DESCRIPTION
**Pull request by dbussink**
This adds functionality so that it's possible to use `Backup` and `BackupShard` and enforce that the backup created is safe to be used for upgrading MySQL.

## Related Issue(s)

Fixes #13448 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/cmd/vtbackup/vtbackup.go [hunk 1]
```diff
@@ -134,7 +134,7 @@ func registerFlags(fs *pflag.FlagSet) {
...
 	fs.BoolVar(&initialBackup, "initial_backup", initialBackup, "Instead of restoring from backup, initialize an empty database with the provided init_db_sql_file and upload a backup of that for the shard, if the shard has no backups yet. This can be used to seed a brand new shard with an initial, empty backup. If any backups already exist for the shard, this will be considered a successful no-op. This can only be done before the shard exists in topology (i.e. before any tablets are deployed).")
 	fs.BoolVar(&allowFirstBackup, "allow_first_backup", allowFirstBackup, "Allow this job to take the first backup of an existing shard.")
-	fs.BoolVar(&restartBeforeBackup, "restart_before_backup", restartBeforeBackup, "Perform a mysqld clean/full restart after applying binlogs, but before taking the backup. Only makes sense to work around xtrabackup bugs.")
+	fs.BoolVar(&restartBeforeBackup, "restart_before_backup", restartBeforeBackup, "Perform a mysqld clean/full restart after applying binlogs, but before taking the backup. This is useful when you want to use this backup for a MySQL upgrade.")
```
Comment (dbussink) [inline go/cmd/vtbackup/vtbackup.go:137]: I've updated this description since it's not just useful for `xtrabackup` bugs but also for MySQL upgrades.

Comment (deepthi) [reply go/cmd/vtbackup/vtbackup.go:137]: We should keep some language around xtrabackup too, instead of deleting it completely.

**Review by deepthi — COMMENTED** [review]
I don't see where we are writing this into the manifest. We should do that too, so that people/automation/code can inspect the manifest to make decisions.

**Comment by dbussink** [comment]
> I don't see where we are writing this into the manifest. We should do that too, so that people/automation/code can inspect the manifest to make decisions.

Let me dig into that. I also then want to make sure that the actual MySQL version is in there as well.

**Comment by dbussink** [comment]
> I don't see where we are writing this into the manifest. We should do that too, so that people/automation/code can inspect the manifest to make decisions.

I have extended the manifest as well. On top of that, if the MySQL version is set in the manifest, `Restore` will now check it as well and print an warning message if a backup can't be used on an older / newer MySQL version.

File: go/cmd/vtbackup/vtbackup.go (hunk 1 above)
Comment (dbussink) [reply go/cmd/vtbackup/vtbackup.go]: Updated this!

**Comment by dbussink** [comment]
> I have extended the manifest as well. On top of that, if the MySQL version is set in the manifest, `Restore` will now check it as well and print an warning message if a backup can't be used on an older / newer MySQL version.

These checks are checks that MySQL itself would also execute, but now we would explicitly catch that before we'd even try to let MySQL restore. So we can catch it earlier this way even.

File: proto/mysqlctl.proto [hunk 2]
```diff
@@ -64,6 +70,7 @@ service MysqlCtl {
   rpc ApplyBinlogFile(ApplyBinlogFileRequest) returns (ApplyBinlogFileResponse) {};
   rpc ReinitConfig(ReinitConfigRequest) returns (ReinitConfigResponse) {};
   rpc RefreshConfig(RefreshConfigRequest) returns (RefreshConfigResponse) {};
+  rpc VersionString(VersionStringRequest) returns (VersionStringResponse) {};
```
Comment (dbussink) [inline proto/mysqlctl.proto:73]: This adds to `mysqlctl`, but that is always upgraded together with `vttablet` which is what uses this API, so this should be all safe to use as well. The only usage is from either `mysqlctl` itself or `vttablet`.

**Review by shlomi-noach — COMMENTED** [review]
Nice touch on adding the MySQL version to the backup manifest. Please see inline comments.

File: go/vt/mysqlctl/backupengine.go [hunk 3]
```diff
@@ -461,14 +483,48 @@ func FindBackupToRestore(ctx context.Context, params RestoreParams, bhs []backup
...
+
+	if flavorTo != flavorFrom {
+		return fmt.Errorf("cannot use backup between different flavors: %q vs. %q", from, to)
+	}
+
+	if parsedTo == parsedFrom {
+		return nil
+	}
+
+	if !parsedTo.atLeast(parsedFrom) {
```
[the remark this answers is not in the document]
Comment (dbussink) [reply go/vt/mysqlctl/backupengine.go:517]: @shlomi-noach That's what it does? Also see the tests. It has a `!` at the start, maybe you missed that? As you said, restoring to 8.0.32 is indeed valid in that case and only if that's `!` the case, we return the error?

File: go/vt/mysqlctl/backupengine.go [hunk 4]
```diff
@@ -461,14 +483,48 @@ func FindBackupToRestore(ctx context.Context, params RestoreParams, bhs []backup
...
-	manifests, err := FindPITRPath(params.RestoreToPos.GTIDSet, manifests)
+	restorePath.manifests, err = FindPITRPath(params.RestoreToPos.GTIDSet, manifests)
 	if err != nil {
 		return nil, err
 	}
-	restorePath.manifests = manifests
 	return restorePath, nil
 }
 
+func isMySQLVersionUpgradeCompatible(to string, from string, upgradeSafe bool) error {
```
[the remark this answers is not in the document]
Comment (dbussink) [reply go/vt/mysqlctl/backupengine.go:493]: Will do!

**Comment by dbussink** [comment]
> @dbussink one thing I've seen with this approach to keep in mind, if you have a large buffer pool and a lot of dirty pages setting `innodb_fast_shutdown` to `0` will take MySQL offline but it will take a long time (sometimes hours) to stop.. Nothing gets logged in the error log in the mean time... so you have no idea what it's doing and how far along it is...

Yeah, so the `vtctld` path that uses `Backup` or `BackupShard` does explicitly shut down MySQL, so that is expected. It could take longer indeed if that happens. 

When using `vtbackup`, I don't think we have this issue since it's a separate / new instance with no sizable dirty buffer pool.

> possible solution for this is to set `innodb_max_dirty_pages_pct` to a low number so MySQL starts aggressively flushing dirty pages to disk, during this time you can monitor `Innodb_buffer_pool_pages_dirty` status variable until it's low enough to issue the shutdown... advantage here is that MySQL keeps running, replication keeps replicating and won't be hours behind after the shutdown...

I think this is more complexity right now than this PR aims to add. It's an explicit opt-in to enable this flag for the builtin backup strategy so nothing that happens without explicitly user knowledge. We could maybe look at this in the future, but I guess it doesn't change the actual runtime, just that we'd know a bit better what is going on?

File: go/vt/mysqlctl/backupengine.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/mysqlctl/backupengine.go:517]: Doh! I missed the `!` !!

**Review by shlomi-noach — APPROVED** [review]

**Review by rohit-nayak-ps — APPROVED** [review]

**Pull request merged**